TinyDRM ILI9486 overlay, and minor bcm2835-codec logging tweak

6by9 opened the pull request:
https://forums.raspberrypi.com/viewtopic.php?t=355484 for the TinyDRM overlay.

bcm2835-codec change for #5582

6by9 committed:
defconfigs: Add TINYDRM_ILI9486 to defconfigs

Adds the DRM driver for the ILI9486.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
dtoverlays: Add drm option to piscreen overlay

Adds the option of selecting the DRM/KMS TinyDRM driver for
this panel, rather than the deprecated FBTFT one.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging: bcm2835-codec: Downgrade the level for a debug message

The debug message from bcm2835_codec_buf_prepare when the buffer
size is incorrect can be a little spammy if the application isn't
careful on how it drives it, therefore drop the priority of the
message.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented on arch/arm/boot/dts/overlays/piscreen-overlay.dts:
```diff
@@ -98,5 +100,7 @@
 		debug =		<&piscreen>,"debug:0";
 		xohms =		<&piscreen_ts>,"ti,x-plate-ohms;0";
+		drm =		<&piscreen>,"compatible=waveshare,rpi-lcd-35",
+				<&piscreen>,"reset-gpios:8=GPIO_ACTIVE_HIGH";
```
Are you sure this works?:
```
$ dtmerge arch/arm/boot/dts/bcm2711-rpi-4-b.dtb base.dtb -
$ dtmerge base.dtb merged.dtb arch/arm/boot/dts/overlays/piscreen.dtbo drm
DTOVERLAY[error]: invalid override value 'GPIO_ACTIVE_HIGH' - ignored
```

6by9 replied:
You're right - my update script to push the change to the Pi obviously didn't for some reason (probably PEBKAC).

6by9 commented:
Actually there's a bigger problem. In the non-drm case (compatible ```ilitek,ili9486```), both drivers get loaded
```
pi@pi:~ $ modinfo fb_ili9486
filename:       /lib/modules/6.1.47-v8+/kernel/drivers/staging/fbtft/fb_ili9486.ko.xz
license:        GPL
author:         Noralf Tronnes
description:    FB driver for the ILI9486 LCD Controller
alias:          platform:ili9486
alias:          spi:ili9486
alias:          platform:fb_ili9486
alias:          spi:fb_ili9486
srcversion:     A87A9FEB7E5C3553E7CDE5A
alias:          of:N*T*Cilitek,ili9486C*
alias:          of:N*T*Cilitek,ili9486
depends:        fbtft
staging:        Y
intree:         Y
name:           fb_ili9486
vermagic:       6.1.47-v8+ SMP preempt mod_unload modversions aarch64

pi@pi:~ $ modinfo ili9486
filename:       /lib/modules/6.1.47-v8+/kernel/drivers/gpu/drm/tiny/ili9486.ko.xz
license:        GPL
author:         Kamlesh Gurudasani <kamlesh.gurudasani@gmail.com>
description:    Ilitek ILI9486 DRM driver
srcversion:     6FBFCC0E15DDF43CA36470A
alias:          of:N*T*Cozzmaker,piscreenC*
alias:          of:N*T*Cozzmaker,piscreen
alias:          of:N*T*Cwaveshare,rpi-lcd-35C*
alias:          of:N*T*Cwaveshare,rpi-lcd-35
alias:          spi:ili9486
depends:        drm_kms_helper,drm,drm_mipi_dbi,backlight,drm_dma_helper
intree:         Y
name:           ili9486
vermagic:       6.1.47-v8+ SMP preempt mod_unload modversions aarch64
```
Presumably it's ```of:N*T*Cilitek,ili9486``` vs ```spi:ili9486``` causing the issue.

pelwell commented:
> Presumably it's of:N*T*Cilitek,ili9486 vs spi:ili9486 causing the issue.

Sorry, missed this pseudo-question - yes, that is probably the cauase.

Can we just disable the fbtft version?

6by9 commented:
> Can we just disable the fbtft version?

Not really unless we change the overlay to only supporting the DRM one.

Both are mainline drivers, so probably worth reporting to mainline to see what they want to do. The fact that SPI has this secondary matching (and complaining if the SPI alias is missing) is really a bit ugly.

pelwell commented:
In the same way that the fbtft driver uses `spi:fb_ili9486`, we could patch the upstream DRM driver so it also uses a prefix for the SPI alias, e.g. `spi:drm_ili9486`.

6by9 commented:
I'm not totally sure on how the spi_id table is used, except that the SPI core whinges if a compatible string device name isn't present in the spi_id table as well. So it complains on ili9486 as ```rpi-lcd-35``` and ```piscreen``` aren't present.
(It also complains that ```ili9486``` is missing for fb_ili9486).

Easiest fix is
```
@ -187,7 +187,8 @@ static const struct of_device_id ili9486_of_match[] = {
 MODULE_DEVICE_TABLE(of, ili9486_of_match);
 
 static const struct spi_device_id ili9486_id[] = {
-       { "ili9486", 0 },
+       { "rpi-lcd-35", 0 },
+       { "piscreen", 0 },
        { }
 };
 MODULE_DEVICE_TABLE(spi, ili9486_id);
```
thus silencing the SPI core code on the mismatch, and removing the clash.

Whilst at it I'll add the missing spi_device_id entries in ads7846 (the touchscreen driver) so at least with the DRM option the SPI core is happy.

6by9 committed:
drm/ili9486: Resolve clash in spi_device_id names

For "Really Good Reasons" [1] the SPI core requires a match
between compatible device strings and the name in spi_device_id.

The ili9486 driver uses compatible strings "waveshare,rpi-lcd-35"
and "ozzmaker,piscreen", but "rpi-lcd-35" and "piscreen" are missing,
so add them.

Compatible string "ilitek,ili9486" is already used by
staging/fbtft/fb_ili9486, therefore leaving it present in ili9486 as an
spi_device_id causes the incorrect module to be loaded, therefore remove
this id.

[1] https://elixir.bootlin.com/linux/latest/source/drivers/spi/spi.c#L487

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
input: ads7846: Add missing spi_device_id strings

The SPI core logs error messages if a compatible string device
name is not also present as an spi_device_id.

No spi_device_id values are specified by the driver, therefore
we get 4 log lines every time it is loaded:
SPI driver ads7846 has no spi_device_id for ti,tsc2046
SPI driver ads7846 has no spi_device_id for ti,ads7843
SPI driver ads7846 has no spi_device_id for ti,ads7845
SPI driver ads7846 has no spi_device_id for ti,ads7873

Add the spi_device_id values for these devices.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented:
This also works, if you prefer:
```
<&piscreen>,"reset-gpios:8=",<GPIO_ACTIVE_HIGH>;
```
but I'm happy to merge as is.

6by9 commented:
> This also works, if you prefer:

Give me two mins to update to that - it's much nicer.
